rosidl_generator_cpp/resource/msg__traits.hpp.template

```diff
@@ -0,0 +1,79 @@
+// generated from rosidl_generator_cpp/resource/msg__traits.hpp.template
+
+@#######################################################################
+@# EmPy template for generating <msg>__traits.hpp files
+@#
+@# Context:
+@#  - spec (rosidl_parser.MessageSpecification)
+@#    Parsed specification of the .msg file
+@#  - subfolder (string)
+@#    The subfolder / subnamespace of the message
+@#    Either 'msg' or 'srv'
+@#  - get_header_filename_from_msg_name (function)
+@#######################################################################
+@
+@{
+from rosidl_generator_cpp import MSG_TYPE_TO_CPP
+
+cpp_namespace = '%s::%s::' % (spec.base_type.pkg_name, subfolder)
+
+cpp_fixed_types = ['bool', 'char', 'uint8_t', 'int8_t', 'uint16_t', 'int16_t', 'uint32_t', 'int32_t', 'uint64_t', 'int64_t', 'float', 'double']
+}@
+#include <stdint.h>
+#include <array>
+#include <string>
+#include <vector>
+
+#ifndef __ALL_MSG__BASIC_TEMPLATES
+#define __ALL_MSG__BASIC_TEMPLATES
+
+template<typename T>
+struct has_fixed_size {};
+
+@[for fixed_type in cpp_fixed_types]@
+template<>
+struct has_fixed_size<@(fixed_type)>
+{
+  static const bool value = true;
+};
+@[end for]@
+
+#endif  /* __ALL_MSG__BASIC_TEMPLATES */
+
+#ifndef __@(spec.base_type.pkg_name)__@(subfolder)__@(get_header_filename_from_msg_name(spec.base_type.type))__TEMPLATES
+#define __@(spec.base_type.pkg_name)__@(subfolder)__@(get_header_filename_from_msg_name(spec.base_type.type))__TEMPLATES
+
+#include "@(spec.base_type.pkg_name)/@(subfolder)/@(get_header_filename_from_msg_name(spec.base_type.type))__struct.hpp"
+
+@{
+fixed_template_string=""
+if not spec.fields:
+    fixed_template_string = "true"
+else:
+    fixed = True
+    for field in spec.fields:
+        if field.type.type == 'string':
+            fixed = False
+            break
+        elif field.type.is_array and (field.type.is_upper_bound or field.type.array_size is None):
+            fixed = False
+            break
+        elif field.type.is_primitive_type():
+            fixed_template_string += "has_fixed_size<{}>::value && ".format(MSG_TYPE_TO_CPP[field.type.type])
+        else:
+            fixed_template_string += "has_fixed_size<{}::msg::{}>::value && ".format(field.type.pkg_name, field.type.type)
+    if not fixed:
+        fixed_template_string = "false"
+    else:
+        fixed_template_string = fixed_template_string[:len(fixed_template_string)-4]
```

Comment:
Can I recommend instead collecting the fixed template strings in a list and then joining them in this case? Something like this:

```python
fixed_template_strings = []
fixed_template_strings.append("...".format(...))
...
fixed_template_string = ' && '.join(fixed_template_strings) if fixed_template_strings else 'true'
```